nuxeo-core/nuxeo-core-io/src/main/java/org/nuxeo/ecm/core/io/download/DownloadServiceImpl.java

```diff
@@ -219,6 +227,109 @@ public String getDownloadUrl(String storeKey) {
         return DownloadService.NXBIGBLOB + "/" + storeKey;
     }
 
+    /**
+     * Gets the base URL of a URL to use to download blobs
+     *
+     * @param urlString the URL to use to download blobs
+     *            ("http://localhost:8080/nuxeo/nxfile/default/3727ef6b-cf8c-4f27-ab2c-79de0171a2c8/files:files/0/file/image.png")
+     * @return the base URL with the context ("http://localhost:8080/nuxeo")
+     * @since 9.1
+     */
+    protected String getBaseDownloadUrl(String urlString) {
+        try {
+            URL url = new URL(urlString);
+            return url.getProtocol() + "://" + url.getHost() + (url.getPort() == -1 ? "" : ":" + url.getPort())
+                    + url.getFile().substring(0, url.getFile().indexOf('/', 1));
+        } catch (MalformedURLException e) {
+            return null;
+        }
+    }
+
+    @Override
+    public Pair<String, String> getDownloadPathAndActionFromUri(String uri) {
+        int slash = uri.indexOf('/');
+        if (slash < 0) {
+            return null;
+        }
+        String type = uri.substring(0, slash);
+        String path = uri.substring(slash + 1);
+        switch (type) {
+        case NXDOWNLOADINFO:
+            // used by nxdropout.js
+            return Pair.of(path, INFO);
+        case NXFILE:
+        case NXBIGFILE:
+            return Pair.of(path, EVENT_NAME_FROM_DOC);
+        case NXBIGZIPFILE:
+        case NXBIGBLOB:
+            return Pair.of(type, EVENT_NAME);
+        default:
+            return null;
+        }
+    }
+
+    @Override
+    public Map<String, String> parseDownloadPath(String path) {
+        Map<String, String> parsedUrl = new HashMap<String, String>();
+        String[] parts = path.split("/");
+        int length = parts.length;
+        if (length < 2) {
+            return null;
+        }
+        String repositoryName = parts[0];
+        String docId = parts[1];
+        String xpath, filename;
+        if (length == 2) {
+            xpath = DownloadService.BLOBHOLDER_0;
+            filename = null;
+        } else if (length == 3) {
+            xpath = parts[2];
+            filename = null;
+        } else {
+            xpath = StringUtils.join(Arrays.asList(parts).subList(2, length - 1), "/");
+            filename = parts[length - 1];
+        }
+        parsedUrl.put("repository", repositoryName);
+        parsedUrl.put("docId", docId);
+        parsedUrl.put("xpath", xpath);
+        parsedUrl.put("filename", filename);
+        return parsedUrl;
+    }
+
+    @Override
+    public Blob resolveBlobFromDownloadUrl(String url) {
+        String baseUrl = getBaseDownloadUrl(url);
+        if (!baseUrl.equals(Framework.getProperty("nuxeo.url"))) {
```

Comment:
In `getBaseDownloadUrl` you can't really identify what the context is. Instead you should just do a `url.startsWith(Framework.getProperty("nuxeo.url"))` I think, and drop `getBaseDownloadUrl`.